core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -0,0 +1,705 @@
+            return new HashDistributionTaskSource(
+                    splitSources,
+                    getPartitionedExchangeSourceHandles(fragment, exchangeSourceHandles),
+                    getReplicatedExchangeSourceHandles(fragment, exchangeSourceHandles),
+                    splitBatchSize,
+                    getSplitTimeRecorder,
+                    bucketToPartitionMap,
+                    bucketNodeMap,
+                    fragment.getPartitioning().getConnectorId());
+        }
+
+        public HashDistributionTaskSource(
+                Map<PlanNodeId, SplitSource> splitSources,
+                Multimap<PlanNodeId, ExchangeSourceHandle> partitionedExchangeSourceHandles,
+                Multimap<PlanNodeId, ExchangeSourceHandle> replicatedExchangeSourceHandles,
+                int splitBatchSize,
+                LongConsumer getSplitTimeRecorder,
+                Optional<int[]> bucketToPartitionMap,
+                Optional<BucketNodeMap> bucketNodeMap,
+                Optional<CatalogName> catalogRequirement)
+        {
+            this.splitSources = ImmutableMap.copyOf(requireNonNull(splitSources, "splitSources is null"));
+            this.partitionedExchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(partitionedExchangeSourceHandles, "partitionedExchangeSourceHandles is null"));
+            this.replicatedExchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(replicatedExchangeSourceHandles, "replicatedExchangeSourceHandles is null"));
+            this.splitBatchSize = splitBatchSize;
+            this.getSplitTimeRecorder = requireNonNull(getSplitTimeRecorder, "getSplitTimeRecorder is null");
+            this.bucketToPartitionMap = requireNonNull(bucketToPartitionMap, "bucketToPartitionMap is null");
+            checkArgument(bucketToPartitionMap.isPresent() || partitionedExchangeSourceHandles.isEmpty() || splitSources.isEmpty(), "bucket to partition map is expected to be set when partitioned inputs are present");
+            this.bucketNodeMap = requireNonNull(bucketNodeMap, "bucketNodeMap is null");
+            checkArgument(splitSources.isEmpty() || bucketNodeMap.isPresent(), "splitToBucketFunction must be set when split sources are present");
+            this.catalogRequirement = requireNonNull(catalogRequirement, "catalogRequirement is null");
+        }
+
+        @Override
+        public List<TaskDescriptor> getMoreTasks()
+        {
+            if (finished || closed) {
+                return ImmutableList.of();
+            }
+
+            Map<Integer, Multimap<PlanNodeId, Split>> partitionToSplitsMap = new HashMap<>();
+            Map<Integer, HostAddress> partitionToNodeMap = new HashMap<>();
+            if (!splitSources.isEmpty()) {
+                for (Map.Entry<PlanNodeId, SplitSource> entry : splitSources.entrySet()) {
+                    PlanNodeId scanNodeId = entry.getKey();
```

Comment:
Technically it does not have to be TableScan node. Maybe just `sourceNodeId`?

Reply:
Renamed to `nodeId`

Reply:
Actually inlined (it was used only once)